src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs

```diff
@@ -475,6 +488,129 @@ async Task<long> GetLengthAsync(CompressionLevel compressionLevel)
+                                    break;
+
+                                case TestScenario.Read:
+                                    while (ZipFileTestBase.ReadAllBytes(s, buffer, 0, buffer.Length) != 0) { };
+                                    break;
+
+                                case TestScenario.ReadAsync:
+                                    while (await ZipFileTestBase.ReadAllBytesAsync(s, buffer, 0, buffer.Length) != 0) { };
+                                    break;
+
+                                case TestScenario.ReadByte:
+                                    while (s.ReadByte() != -1) { }
+                                    break;
+
+                                case TestScenario.ReadByteAsync:
+                                    while (await s.ReadByteAsync() != -1) { }
+                                    break;
+                            }
+                        }
+                        catch (InvalidDataException e)
+                        {
+                            if (expectException)
+                                continue;
+
+                            throw new XunitException($"An unexpected error occured while decompressing data:{e}");
+                        }
+
+                        if (expectException)
+                        {
+                            throw new XunitException($"Truncated stream was decompressed successfully but exception was expected: length={i}/{compressedData.Length}");
+                        }
+                    }
+                }
+            }
+        }
+
+        [Fact]
+        public void StreamCorruption_IsDetected()
+        {
+            var source = Enumerable.Range(0, 64).Select(i => (byte)i).ToArray();
+            var buffer = new byte[64];
+            byte[] compressedData;
+            using (var compressed = new MemoryStream())
+            using (var gzip = new GZipStream(compressed, CompressionMode.Compress))
+            {
+                foreach (var b in source)
+                {
+                    gzip.WriteByte(b);
+                }
+
+                gzip.Dispose();
+                compressedData = compressed.ToArray();
+            }
+
+            for (var byteToCorrupt = 0; byteToCorrupt < compressedData.Length; byteToCorrupt++)
+            {
+                // corrupt the data
+                compressedData[byteToCorrupt]++;
+
+                using (var decompressedStream = new MemoryStream(compressedData))
+                {
+                    using (var s = new GZipStream(decompressedStream, CompressionMode.Decompress))
+                    {
+                        try
+                        {
+                            while(ZipFileTestBase.ReadAllBytes(s, buffer, 0, buffer.Length) != 0) { };
+
+                            Assert.Equal(source, buffer);
+                        }
+                        catch (InvalidDataException)
+                        {
+                        }
```

Comment:
This will eat-up false positives. You can do this instead:
```cs 
Assert.Throws<InvalidDataException>(() => { while (ZipFileTestBase.ReadAllBytes(s, buffer, 0, buffer.Length) != 0) ; });
```

If there are cases where a corruption can't be expected (e.g: changing a header byte), we should skip them, but they should be deliberated.

Reply:
Ok, I did this (code may not be elegant though, let me know if you have a more elegant version in mind).

Skipping though means that we do not check the decompression process was successful in these cases (e.g. comparing the decompressed buffer with the initial source). It may or may not be relevant to do so, I'm not sure.

Reply:
I'm not exactly sure what you mean here, but I think the way you have it in the latest commits is the way to go, we should always assert for the exception when desired, not just ignore it (unless there's a reason for it).
https://github.com/dotnet/runtime/blob/e4c1d7fdbbc2a9c8bb9667cc9c715291799dbd14/src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.ZLib.cs#L49-L52